### PR TITLE
feat/team-crud-v2

### LINKED ISSUE #148: User Story: Team CRUD V2
As a Team owner
I want to create different team
When i CRUD team
Then i got Team CRUD Where i’m the Team owner & team member

### PR DESCRIPTION
**Pull request by dasarathg68**
# Description

As a Team owner
I want to create different team
When i CRUD team
Then i get Team CRUD Where i’m the Team owner & team member

Fixes #148 #147 

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

* **Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: backend/src/controllers/teamController.ts [hunk 1]
```diff
@@ -52,46 +79,87 @@ const getTeam = async (req: Request, res: Response) => {
...
-// update team
+// Update team
 const updateTeam = async (req: Request, res: Response) => {
-  /* 
+  /*
   #swagger.tags = ['Teams']
   */
   const { id } = req.params;
   const { name, description, bankAddress } = req.body;
+  const callerAddress = req.headers.calleraddress;
```
Comment (hermannleboss) [inline backend/src/controllers/teamController.ts:149]: Why you get the caller address in the header ?

File: backend/src/controllers/teamController.ts [hunk 2]
```diff
@@ -52,46 +79,87 @@ const getTeam = async (req: Request, res: Response) => {
...
+    const team = await prisma.team.findUnique({
+      where: {
+        id: Number(id),
+      },
+    });
+    if (!team) {
+      return errorResponse(404, "Team not found", res);
+    }
+    if (team.ownerAddress !== callerAddress) {
+      return errorResponse(401, "Unauthorized", res);
```
Comment (hermannleboss) [inline backend/src/controllers/teamController.ts:160]: Should use 403 Error it's more clear.

File: backend/src/controllers/userController.ts
Comment (hermannleboss) [inline backend/src/controllers/userController.ts:1]: You should also add an endpoint to get All Users with pagination.

File: backend/src/routes/teamRoutes.ts
Comment (hermannleboss) [inline backend/src/routes/teamRoutes.ts:1]: I like the way you chose /:id/member for the endpoint =

File: backend/src/routes/userRoutes.ts [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
+  getUser,
+  updateUser,
+  searchUser,
+} from "../controllers/userController";
 const userRoutes = express.Router();
 
 userRoutes.get("/nonce/:address", getNonce);
 userRoutes.get("/:address", getUser);
 userRoutes.put("/:address", updateUser);
+userRoutes.get("/search/user", searchUser);
```
Comment (hermannleboss) [inline backend/src/routes/userRoutes.ts:13]: Here no need to add user at the end. 
the final route will be /api/user/search it's more simple

**Review by hermannleboss — CHANGES_REQUESTED** [review]

File: backend/src/routes/userRoutes.ts (hunk 3 above)
Comment (dasarathg68) [reply backend/src/routes/userRoutes.ts:13]: The reason I added it is because the routing somehow gets ambiguous and calls the`userRoutes.get("/:address", getUser);`instead

File: backend/src/controllers/teamController.ts (hunk 1 above)
Comment (dasarathg68) [reply backend/src/controllers/teamController.ts:149]: Because I wanted to set a standard for all requests. Normally in GET and DELETE operations, we wouldn't be able to send a body. That's why I followed this convention of getting callerAddress in the headers.

File: backend/src/routes/userRoutes.ts (hunk 3 above)
Comment (hermannleboss) [reply backend/src/routes/userRoutes.ts:13]: User Routes aren't prefixed by user ?

like /api/user/:address ?

File: backend/src/controllers/teamController.ts [hunk 4]
```diff
@@ -52,46 +79,87 @@ const getTeam = async (req: Request, res: Response) => {
...
-        ownerId: ownerId,
+        ownerAddress: callerAddress,
+      },
+      include: {
+        members: true,
       },
     });
-    res.status(200).json({ teams, success: true });
-  } catch (error) {
-    return errorResponse(500, error, res);
```
[the remark this answers is not in the document]
Comment (dasarathg68) [reply backend/src/controllers/teamController.ts]: Ah yes, nice catch!
Changed it now.

File: backend/src/routes/userRoutes.ts (hunk 3 above)
Comment (dasarathg68) [reply backend/src/routes/userRoutes.ts:13]: it is, but the thing is /api/user/:address and /api/user/search is creating ambiguity for the express router.
/api/user/:address: The :address part is a dynamic segment, meaning it can match any value in that position.
/api/user/search: This is a static path meant to match exactly the word "search".

So the dynamic path takes precedence and proceeds to evaluate the path params as "search"

File: backend/src/routes/teamRoutes.ts
Comment (dasarathg68) [reply backend/src/routes/teamRoutes.ts:1]: Thanks!

**Comment by dasarathg68** [comment]
@hermannleboss implemented pagination and also followed the other recommendations. Please check.

**Review by hermannleboss — APPROVED** [review]

File: backend/src/routes/userRoutes.ts (hunk 3 above)
Comment (hermannleboss) [reply backend/src/routes/userRoutes.ts:14]: I think the order matter

Comment (hermannleboss) [reply backend/src/routes/userRoutes.ts:14]: You should put /search before address

Comment (dasarathg68) [reply backend/src/routes/userRoutes.ts:14]: Ah works! Thanks.

File: backend/src/controllers/teamController.ts (hunk 1 above)
Comment (dasarathg68) [reply backend/src/controllers/teamController.ts:135]: I'm not sure about how to do this

**Comment by hermannleboss** [comment]
I just remark you don't add too much test.

**Comment by dasarathg68** [comment]
> I just remark you don't add too much test.

I'm still studying about the best practices in testing. 
For eg., farrell just sent me this article
 https://github.com/goldbergyoni/javascript-testing-best-practices?tab=readme-ov-file#-%EF%B8%8F27-avoid-global-test-fixtures-and-seeds-add-data-per-test

TLDR: using seed script globally messes the tests up.
So, I will push the tests in some time after studying these practices and the links u sent.
